vendor/github.com/kubernetes-csi/external-snapshotter/client/v3/apis/volumesnapshot/v1/types.go

```diff
@@ -0,0 +1,395 @@
+
+	// spec defines properties of a VolumeSnapshotContent created by the underlying storage system.
+	// Required.
+	Spec VolumeSnapshotContentSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+
+	// status represents the current information of a snapshot.
+	// +optional
+	Status *VolumeSnapshotContentStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// VolumeSnapshotContentList is a list of VolumeSnapshotContent objects
+// +kubebuilder:object:root=true
+type VolumeSnapshotContentList struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// items is the list of VolumeSnapshotContents
+	Items []VolumeSnapshotContent `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// VolumeSnapshotContentSpec is the specification of a VolumeSnapshotContent
+type VolumeSnapshotContentSpec struct {
+	// volumeSnapshotRef specifies the VolumeSnapshot object to which this
+	// VolumeSnapshotContent object is bound.
+	// VolumeSnapshot.Spec.VolumeSnapshotContentName field must reference to
+	// this VolumeSnapshotContent's name for the bidirectional binding to be valid.
+	// For a pre-existing VolumeSnapshotContent object, name and namespace of the
+	// VolumeSnapshot object MUST be provided for binding to happen.
+	// This field is immutable after creation.
+	// Required.
+	VolumeSnapshotRef core_v1.ObjectReference `json:"volumeSnapshotRef" protobuf:"bytes,1,opt,name=volumeSnapshotRef"`
+
+	// deletionPolicy determines whether this VolumeSnapshotContent and its physical snapshot on
+	// the underlying storage system should be deleted when its bound VolumeSnapshot is deleted.
+	// Supported values are "Retain" and "Delete".
+	// "Retain" means that the VolumeSnapshotContent and its physical snapshot on underlying storage system are kept.
+	// "Delete" means that the VolumeSnapshotContent and its physical snapshot on underlying storage system are deleted.
+	// In dynamic snapshot creation case, this field will be filled in with the "DeletionPolicy" field defined in the
+	// VolumeSnapshotClass the VolumeSnapshot refers to.
+	// For pre-existing snapshots, users MUST specify this field when creating the VolumeSnapshotContent object.
+	// Required.
+	DeletionPolicy DeletionPolicy `json:"deletionPolicy" protobuf:"bytes,2,opt,name=deletionPolicy"`
+
+	// driver is the name of the CSI driver used to create the physical snapshot on
+	// the underlying storage system.
+	// This MUST be the same as the name returned by the CSI GetPluginName() call for
+	// that driver.
+	// Required.
+	Driver string `json:"driver" protobuf:"bytes,3,opt,name=driver"`
+
+	// name of the VolumeSnapshotClass to which this snapshot belongs.
+	// +optional
+	VolumeSnapshotClassName *string `json:"volumeSnapshotClassName,omitempty" protobuf:"bytes,4,opt,name=volumeSnapshotClassName"`
```

Comment:
Should we create a Parameters field and copy the `SnapshotClass.parameters` field here. A snapshot must be able to operate independently of the `SnapshotClass` after provisioning. We have `DeletionPolicy` here already, which is good, but do we need parameters too?

Reply:
@msau42 reminded me that PV/PVC actually handles this in a better way. Instead of copying all parameters over from the Class object (which may include lots of unnecessary information), the CSI `CreateVolumeResponse` contains a set of `volume_context` -- basically letting the driver decide what additional important info should be maintained about the volume. We should do something similar here. See https://github.com/container-storage-interface/spec/blob/master/csi.proto#L488

Reply:
We'll take a look.